### PR TITLE
8290917: x86: Memory-operand arithmetic instructions have too low costs

### PR DESCRIPTION
**Pull request by merykitty**
The pattern `AddI (LoadI mem) imm` should be matched by a load followed by an add with constant, instead, it is currently matched as a constant load followed by an add with memory. The reason is that the cost of `addI_rReg_mem` is too low, this patch fixes this by increasing the cost of this fused instruction.

Testing: Manually run the test case in the JBS and look at the compiled code.

I also do some small clean-ups in x86_64.ad:

- The `mulHiL` rules have unnecessary constraints on the input registers, these can be removed. The `no_rax_RegL` operand as a consequence can also be removed.
- The rules involving long division by a constant can be removed because it has been covered by the optimiser during idealisation.
- The pattern `SubI src imm` and the likes never match because they are converted to `AddI src -imm` by the optimiser. As a result, these rules can be removed
- The rules involving shifting the argument by 1 are covered by and exactly the same as the corresponding rules of shifting by an immediate. As a result, they can be removed.
- Some rules involving and-ing with a bit mask have unnecessary constraints on the target register.

Please kindly review, thank you very much.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290917](https://bugs.openjdk.org/browse/JDK-8290917): x86: Memory-operand arithmetic instructions have too low costs

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to [ef4f3cf9](https://git.openjdk.org/jdk/pull/9791/files/ef4f3cf91e909b4508cff3c3f9728860aed1cd16)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9791/head:pull/9791` \
`$ git checkout pull/9791`

Update a local copy of the PR: \
`$ git checkout pull/9791` \
`$ git pull https://git.openjdk.org/jdk pull/9791/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9791`

View PR using the GUI difftool: \
`$ git pr show -t 9791`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9791.diff">https://git.openjdk.org/jdk/pull/9791.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
`instruct incI_mem(memory dst, immI_1 src, rFlagsReg cr)` should be one instruction, not 2 as you pointed in JBS, because it increments by 1. The same should be true for other instructions which has `immI_1`, `immL1`, `immI_M1`, `immL_M1`.

I agree with your removal of `Sub*` instructions.

Would be nice to have JMH micros for all these changes to verify that score stay the same or better.

**Comment by merykitty** [comment]
Thanks for your review, the pattern mentioned in the JBS is to load from memory, do an addition with a constant and have the result as a local variable, which is different from `incI_mem` which requires the result to be stored back to the exact same memory.

`incI_mem` is removed because the pattern has already been covered by `addI_mem_imm` and the latter produces better code. Similarly, the pattern covered by `salI_rReg_1` is already covered by `salI_rReg_imm` and these 2 rules otherwise are exactly the same, which is redundant.

I have added a JMH benchmark for the affected cases and the results are as follow:

                                            Before             After
    Benchmark               Mode  Cnt    Score   Error     Score   Error  Units   Change
    BasicRules.add_mem_con  avgt    5  200.078 ± 1.555   200.663 ± 1.919  ns/op    +0.3%
    BasicRules.inc_mem      avgt    5  355.942 ± 0.695   322.630 ± 1.740  ns/op    -9.4%

**Comment by TobiHartmann** [comment]
Intel folks (@jatin-bhateja, @sviswa7?) should have a look at this as well.

**Comment by sviswa7** [comment]
I don't think we should replace the inc/dec by add.

On my desktop, I see the following:
Before:
Benchmark               Mode  Cnt    Score   Error  Units
BasicRules.add_mem_con  avgt    3  132.268 ± 0.599  ns/op
BasicRules.inc_mem      avgt    3  169.980 ± 0.617  ns/op

After:
Benchmark               Mode  Cnt    Score   Error  Units
BasicRules.add_mem_con  avgt    3  117.426 ± 0.128  ns/op
BasicRules.inc_mem      avgt    3  182.907 ± 0.277  ns/op

The inc_mem jmh performance is worse after the patch. 

There is already UseIncDec option which is set appropriately to select whether to generate inc/dec or the add/sub instruction.

**Comment by merykitty** [comment]
@sviswa7 Thanks a lot for your review, I have reverted that change. I don't understand why, though, it does not seem that the bottleneck is in the predecoder.

**Comment by sviswa7** [comment]
@merykitty Thanks for reverting those changes. Could you please also add jmh tests for the following:
 1) AndL with 255
 2) AndL with 65535
 3) DivL by 10
For 1) and 2) we are changing the instruction from q version to l version, so want to make sure the performance is at par atleast.
For 3) it will be good to check that the compiler is optimizing divide by 10 for long data type as well now.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_64.ad:10209]: Good to add a comment here that the upper 32 bits are zeroed by the instruction.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_64.ad:10221]: Good to add a comment here that the upper 32 bits are zeroed by the instruction.

**Comment by merykitty** [comment]
@sviswa7 Thanks for your reviews, I have addressed those in the last commits, the benchmark results are as follow:

                                                    Before            After
    Benchmark                      Mode  Cnt    Score    Error    Score    Error  Units
    BasicRules.add_mem_con         avgt   15  203.470 ±  8.955  202.771 ±  2.867  ns/op
    BasicRules.andL_rReg_imm255    avgt   15  183.892 ±  2.314  185.008 ±  2.642  ns/op
    BasicRules.andL_rReg_imm65535  avgt   15  183.854 ±  2.293  184.849 ±  2.668  ns/op
    BasicRules.divL_10             avgt   15  643.210 ± 10.536  645.040 ± 10.014  ns/op

**Review by vnkozlov — COMMENTED** [review]
Please also add testing (JMH) for shifts by 1 which you removed. And for `sub_*_imm` instructions.

**Review by sviswa7 — APPROVED** [review]
Looks good to me. Please wait for Vladimir's review and approval.

**Comment by merykitty** [comment]
@vnkozlov I have added cases for shifts by 1 and immediate subtractions as you suggested.

                                                   Before             After
    Benchmark                      Mode  Cnt    Score    Error    Score    Error  Units
    BasicRules.add_mem_con         avgt   15  202.913 ±  4.915  210.006 ±  9.775  ns/op
    BasicRules.andL_rReg_imm255    avgt   15  185.073 ±  2.768  187.580 ±  4.930  ns/op
    BasicRules.andL_rReg_imm65535  avgt   15  184.056 ±  2.315  186.884 ±  2.198  ns/op
    BasicRules.divL_10             avgt   15  644.609 ±  8.740  654.411 ± 11.558  ns/op
    BasicRules.salI_rReg_1         avgt   15  183.385 ±  2.720  181.688 ±  1.435  ns/op
    BasicRules.salL_rReg_1         avgt   15  183.192 ±  2.597  184.337 ±  7.378  ns/op
    BasicRules.sarI_rReg_1         avgt   15  184.174 ±  2.646  184.984 ±  6.895  ns/op
    BasicRules.sarL_rReg_1         avgt   15  182.288 ±  2.012  182.062 ±  2.178  ns/op
    BasicRules.shrI_rReg_1         avgt   15  182.465 ±  2.242  180.915 ±  1.162  ns/op
    BasicRules.shrL_rReg_1         avgt   15  183.147 ±  2.277  181.170 ±  1.567  ns/op
    BasicRules.subI_rReg_imm       avgt   15  200.593 ±  2.279  199.828 ±  2.255  ns/op
    BasicRules.subL_rReg_imm       avgt   15  185.141 ±  5.717  183.781 ±  3.173  ns/op

File: src/hotspot/cpu/x86/x86_64.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86_64.ad:10210]: Saving of redundant REX byte is allocations happen in lower register bank look goods.

**Review by jatin-bhateja — APPROVED** [review]
Patch is doing some paper cuts optimization, looks ok to me.

**Review by vnkozlov — COMMENTED** [review]
Thank you for updating tests. I am running testing now.

**Review by vnkozlov — APPROVED** [review]
My testing passed.

**Comment by merykitty** [comment]
Thanks very much for your reviews!
/integrate

**Pull request closed without merging**
